sys/include/chunked_ringbuffer.h

```diff
@@ -0,0 +1,224 @@
+
+/**
+ * @brief Close the current chunk
+ *
+ * @param[in] rb        The Ringbuffer to work on
+ * @param[in] valid     True if the chunk is valid and should be stored
+ *                      False if the current chunk should be discarded
+ *
+ * @return true         If the chunk could be stored in the valid chunk array
+ * @return false        If there is no more space in the valid chunk array
+ */
+bool crb_end_chunk(chunk_ringbuf_t *rb, bool valid);
+
+/**
+ * @brief Add a complete chunk to the Ringbuffer
+ *
+ * This is a convenience function that combines @ref crb_start_chunk,
+ * @ref crb_add_bytes and @ref crb_end_chunk
+ *
+ * @param[in] rb        The Ringbuffer to work on
+ * @param[in] data      The data to write
+ * @param[in] len       Size of data
+ *
+ * @return true         If the chunk could be added to the valid chunk array
+ * @return false        There was not enough space and the chunk was discarded
+ */
+bool crb_add_chunk(chunk_ringbuf_t *rb, const void *data, size_t len);
+
+/**
+ * @brief Get the size of the first valid chunk
+ *
+ * @param[in] rb        The Ringbuffer to work on
+ * @param[out] len      Pointer to store the size of the first valid chunk
+ *
+ * @return true         If a valid chunk exists and @p size was written
+ * @return false        If no valid chunk exists
+ */
+bool crb_get_chunk_size(chunk_ringbuf_t *rb, size_t *len);
+
+/**
+ * @brief Get a number of bytes from the first valid chunk
+ *
+ * @param[in] rb        The Ringbuffer to work on
+ * @param[out] dst      Destination buffer
+ * @param[in] offset    Offset to the start of the chunk
+ * @param[in] len       Number of bytes to read
+ *
+ * @return true         If the data could be read
+ * @return false        If no valid chunk exists or the bytes could not be read
+ */
+bool crb_peek_bytes(chunk_ringbuf_t *rb, void *dst, size_t offset, size_t len);
+
+/**
+ * @brief Remove a chunk from the valid chunk array
+ *
+ * @param[in] rb        The Ringbuffer to work on
+ * @param[out] dst      Destination where the chunk contents should be copied to.
+ *                      May be NULL, then the chunk is just discarded.
+ * @param[in] len       Max number of bytes to read. If there are bytes left in the
+ *                      chunk beyond that, they will be discarded
+ *
+ * @return true         If a chunk was consumed
+ * @return false        If no valid chunk did exist
+ */
+bool crb_consume_chunk(chunk_ringbuf_t *rb, void *dst, size_t len);
+
+/**
+ * @brief Execute a callback for each byte in the first valid chunk
+ *
+ * @param[in] rb        The Ringbuffer to work on
+ * @param[in] func      The function to call for each byte
+ * @param[in] ctx       Optional function argument
+ *
+ * @return true         If a valid chunk exits on which the function was executed
+ * @return false        If no valid chunk exists
+ */
+bool crb_chunk_foreach(chunk_ringbuf_t *rb, crb_byte_callback_t func, void *ctx);
```

Comment:
Any particular reason to do this per-byte and not per-slice?

Reply:
That's the part where my generalization was at risk of breaking down:  DOSE checks if the CRC of the received frame is valid before emitting a RX event. That means we don't have a buffer yet into which the data has been copied, so we need to calculate the CRC on the ringbuffer.

An alternative would be do calculate the CRC on-line (on each RX byte as it comes in) and store the result inside `dose_t`.
I decided against that because my time budget inside the ISR is already pretty tight, so I didn't want to add more calculations - I did not benchmark this though.

So I added a generic function to perform any operation on the bytes of the chunk, which just happens to be CRC in this case. (And will end up costing more instructions this way than doing it on-line, but with less of a deadline)

Reply:
I don't quite understand how this affects CRC calculation. If the callback was called 1 or 2 times (with and length) vs. 100 times, either happens outside of the critical path, but the former can can be optimized better.

Reply:
Ah now I see where you are getting at!
Yes, will do.